### PR TITLE
Feature/python.paddle.v2

### PR DESCRIPTION
**Pull request by reyoung**
Rearrange Paddle Packages to `paddle.v2`. Use `import paddle.v2 as paddle` currently.

### CODE REVIEW

File: demo/mnist/api_train.py [hunk 1]
```diff
@@ -6,42 +6,33 @@
...
 The user api could be simpler and carefully designed.
 """
-import py_paddle.swig_paddle as api
-from py_paddle import DataProviderConverter
-import paddle.trainer.PyDataProvider2 as dp
-import numpy as np
 import random
+
+import paddle.v2 as paddle
+import py_paddle.swig_paddle as api
```
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: 我总觉得这个swig_paddle需要稍微封装一下

File: demo/mnist/api_train.py [hunk 2]
```diff
@@ -6,42 +6,33 @@
...
 
 
 def network_config():
-    imgs = data_layer(name='pixel', size=784)
-    hidden1 = fc_layer(input=imgs, size=200)
-    hidden2 = fc_layer(input=hidden1, size=200)
-    inference = fc_layer(input=hidden2, size=10, act=SoftmaxActivation())
-    cost = classification_cost(
-        input=inference, label=data_layer(
+    imgs = paddle.config.data_layer(name='pixel', size=784)
```
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: network的config和上边那些optimizer的config是不是分开比较好

File: demo/mnist/api_train.py (hunk 1 above)
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: 看到下边已经改了

File: python/setup.py.in [hunk 3]
```diff
@@ -4,7 +4,8 @@ packages=['paddle',
           'paddle.proto',
           'paddle.trainer',
           'paddle.trainer_config_helpers',
-          'paddle.utils']
+          'paddle.utils',
+	  'paddle.v2']
```
Comment (jacquesqiao) [inline python/setup.py.in]: 貌似有个缩进的问题

File: demo/mnist/api_train.py (hunk 1 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: done.

File: python/setup.py.in (hunk 3 above)
Comment (reyoung) [inline python/setup.py.in]: done.

File: demo/mnist/api_train.py (hunk 2 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 其实都是生成protobuf的东西。先放到一个包里面了。

**Comment by jacquesqiao** [comment]
lgtm

File: demo/mnist/api_train.py [hunk 4]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
     # enable_types = [value, gradient, momentum, etc]
     # For each optimizer(SGD, Adam), GradientMachine should enable different
     # buffers.
-    opt_config_proto = parse_optimizer_config(optimizer_config)
-    opt_config = api.OptimizationConfig.createFromProto(opt_config_proto)
-    _temp_optimizer_ = api.ParameterOptimizer.create(opt_config)
+    opt_config_proto = paddle.config.parse_optimizer(optimizer_config)
+    opt_config = paddle.raw.OptimizationConfig.createFromProto(opt_config_proto)
+    _temp_optimizer_ = paddle.raw.ParameterOptimizer.create(opt_config)
     enable_types = _temp_optimizer_.getParameterTypes()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 这一段没看明白——我们需要optimizer吗？为什么只create了一个temproary optimzier，然后取了其中一个property（enable_types），用来生成一个 gradient machine 就完了。我们到底需不需要optimizer这个概念呢？

File: demo/mnist/api_train.py [hunk 5]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
     # get enable_types for each optimizer.
     # enable_types = [value, gradient, momentum, etc]
     # For each optimizer(SGD, Adam), GradientMachine should enable different
     # buffers.
-    opt_config_proto = parse_optimizer_config(optimizer_config)
-    opt_config = api.OptimizationConfig.createFromProto(opt_config_proto)
-    _temp_optimizer_ = api.ParameterOptimizer.create(opt_config)
+    opt_config_proto = paddle.config.parse_optimizer(optimizer_config)
+    opt_config = paddle.raw.OptimizationConfig.createFromProto(opt_config_proto)
+    _temp_optimizer_ = paddle.raw.ParameterOptimizer.create(opt_config)
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: L73 到 L75 看上去是要按照根据一些config信息创建一个optimizer？如果是这样为什么需要 protobuf message 和一个高阶函数 parse_optimizer 呢？是不是下面这样就可以了：


```
optimizer = paddle.v2.optimizer.Adam(
    learning_rate=1e-4,
    batch_size=1000,
    model_averager=paddle.v2.config.ModelAverage(average_window=0.5),
    regularizator=paddle.v2.regularizer.L2(rate=0.5))
```

这样也不需要定义 `def optimizer_config()` 这个函数了。

File: demo/mnist/api_train.py [hunk 6]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
     enable_types = _temp_optimizer_.getParameterTypes()
 
     # Create Simple Gradient Machine.
-    model_config = parse_network_config(network_config)
-    m = api.GradientMachine.createFromConfigProto(
-        model_config, api.CREATE_MODE_NORMAL, enable_types)
+    model_config = paddle.config.parse_network(network_config)
+    m = paddle.raw.GradientMachine.createFromConfigProto(
+        model_config, paddle.raw.CREATE_MODE_NORMAL, enable_types)
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 类似的，如果这里只是要create一个gradient mahcine，貌似应该是：

```
images  = paddle.v2.layer.data(name='pixel', size=784)
hidden1 = paddle.v2.layer.fc(input=images, size=200)
hidden2 = paddle.v2.layer.fc(input=hidden1, size=200)
classes = paddle.v2.layer.fc(input=hidden2, size=10, act=paddle.config.SoftmaxActivation())
cost    = paddle.v2.cost.classification(
    input=classes, 
    label=paddle.v2.layer.data(name='label', size=10))
gm = paddle.gradient_machine.create(cost)
```

这样也不需要定义 `network_config`  函数了？

File: demo/mnist/api_train.py [hunk 7]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
 
     # This type check is not useful. Only enable type hint in IDE.
     # Such as PyCharm
-    assert isinstance(m, api.GradientMachine)
+    assert isinstance(m, paddle.raw.GradientMachine)
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 我理解如果 m/gm 是上面一行 `gm = paddle.v2.gradient_machine.create(...)` 这样产生的，读者自然相信 gm 的类型是一个  gradient machine，也就不需要这个 assertation 来增加程序的可读性了。

File: demo/mnist/api_train.py [hunk 8]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
 
     # Initialize Parameter by numpy.
-    init_parameter(network=m)
+    m.randParameters()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: randParameters是一个函数吗？按照[Python style guide](https://www.python.org/dev/peps/pep-0008/#id43)，函数名应该是 `randomize_parameters`。

File: demo/mnist/api_train.py [hunk 9]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
 
     # Create Local Updater. Local means not run in cluster.
     # For a cluster training, here we can change to createRemoteUpdater
     # in future.
-    updater = api.ParameterUpdater.createLocalUpdater(opt_config)
-    assert isinstance(updater, api.ParameterUpdater)
+    updater = paddle.raw.ParameterUpdater.createLocalUpdater(opt_config)
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]:
```
updater = paddle.v2.parameter_updater.create(
    optimizer = paddle.v2.optimizer.Adam(),
    learning_rate=1e-4,
    batch_size=1000,
    model_averager=paddle.v2.config.ModelAverage(average_window=0.5),
    regularizator=paddle.v2.regularizer.L2(rate=0.5))
```

File: demo/mnist/api_train.py [hunk 10]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
+    assert isinstance(updater, paddle.raw.ParameterUpdater)
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 类似的，如果 updater 是上面一行那样create的，读者应该不依赖这一行来了解updater的类型了。

File: demo/mnist/api_train.py [hunk 11]
```diff
@@ -73,42 +63,44 @@ def input_order_converter(generator):
...
 
     # Initialize ParameterUpdater.
     updater.init(m)
 
     # DataProvider Converter is a utility convert Python Object to Paddle C++
     # Input. The input format is as same as Paddle's DataProvider.
-    converter = DataProviderConverter(
-        input_types=[dp.dense_vector(784), dp.integer_value(10)])
+    converter = paddle.data.DataProviderConverter(input_types=[
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 这里为什么需要暴露一个 converter 出来呢？按照上面comment的解释——把一个Python obj转换为一个C++ obj——这个样的逻辑不应该出现在API里，暴露给用户吧？

File: python/paddle/v2/config.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+from paddle.trainer_config_helpers import *
```
Comment (wangkuiyi) [inline python/paddle/v2/config.py]: 这里 import *，让我们丧失了对 paddle.v2 package 里的内容的掌握——只要有人修改了 paddle.trianer_config_helpers 里的内容，这里的symbols就发生变化了吧？

我建议，我们趁此机会，先把 mnist 这个 demo 需要的内容 copy-n-paste 过来。然后依据把 mnist demo 写的让读者能“顾名思义”的原则，修改 copy 过来的库。

随后我们一个一个demo的过，重复上述过程，得到的 paddle.v2 应该就是我们想要的了吧。

File: demo/mnist/api_train.py [hunk 13]
```diff
@@ -178,7 +170,7 @@ def main():
         test_data_generator = input_order_converter(read_from_mnist(test_file))
         for data_batch in generator_to_batch(test_data_generator, 512):
```
Comment (emailweixu) [inline demo/mnist/api_train.py]: 如果能写成下面这样会更容易让用户理解：
data_reader = read_from_mnist(test_file)
for data_batch in data_reader:
  m.forward(data_batch, outArgs, paddle.raw.PASS_TEST)
...

input_order_converter 和 converter这两个东西会让用户难以理解。用户能做的多半只是照抄代码，但是并不知道为什么要用这些converter. 前面创建optimizer的那段代码也是同样的问题。

一个好的API最好能做到每一步调用都让用户很容易理解为什么要做。

File: demo/mnist/api_train.py (hunk 5 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 当然，这样更好些。

File: demo/mnist/api_train.py (hunk 6 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 这样实现，Paddle需要比较大规模的重构。

主要是现在Paddle解析网络配置的过程，是调用fc之类的函数，在这个函数里面写了一个全局的变量。
而传cost的方式，相当于在这个cost变量里，记录了网络的所有拓扑。这需要我们的返回值记录原来全局变量中的信息。

并且，其实还有一个问题，就是有可能一个神经网络会有多个输出值。可能不只有一个cost。使用传函数的办法也并不算非常不优雅。

File: demo/mnist/api_train.py (hunk 7 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 这行完全可以不需要。主要是为IDE增加的type信息。常见的IDE可以根据这个type信息做代码提示。

写这行的主要原因是我们用swig生成的python代码。swig并没有按照Python的风格注释函数返回值的类型。如果我们从C-API写的话，就没这个问题了。

File: demo/mnist/api_train.py (hunk 8 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 是。但这里因为使用的是SWIG编译的C++头文件，所以还是按照C++的命名方式了。如果用C-API暴露，就没有这个问题了。

File: demo/mnist/api_train.py (hunk 9 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 没问题。有道理。

File: demo/mnist/api_train.py (hunk 11 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 是的，可以封装到数据的Iterator里面。

File: demo/mnist/api_train.py (hunk 6 above)
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: 在目前config_parser的实现里边，还必须是一个函数，或者是一个独立的文件，而且这里的改造代价比较大。https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer/config_parser.py#L3444
这是一个几千行代码的文件，后续应该是要重构掉的。

File: demo/mnist/api_train.py (hunk 8 above)
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: randParameters是从cpp文件中通过swig暴露出来的接口

File: python/paddle/v2/config.py (hunk 12 above)
Comment (reyoung) [inline python/paddle/v2/config.py]: 不过其实paddle.trainer_config_helpers里面暴露的符号是严格控制的。里面使用了__all__来控制符号暴露。

copy and paste这个包，会少暴露非常多东西。比如，我们教程里面的MNIST可能使用全连接做的。用户可能想改成卷积之类的操作。但是，如果只是copy and paste demo需要的接口的话，卷积很可能就没复制过去。这用户就缺乏了这部分灵活性了。


同时，和之前的一个comment类似，如果我们真的需要使用『返回值』而不是『函数』来去定义网络结构的话，那其实所有的配置解析都要重写一下。copy and paste反而不好，不如直接重写一个解析过程。

File: demo/mnist/api_train.py (hunk 13 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 有道理。好的。

这两个可以直接patch到forward这个函数里面。

File: demo/mnist/api_train.py (hunk 4 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: 不需要optimizer这个概念。

File: demo/mnist/api_train.py (hunk 6 above)
Comment (reyoung) [inline demo/mnist/api_train.py]: @wangkuiyi 如果确定传变量的方式没有问题的话，那么我们要不就重构一下这个解析了？ @wangkuiyi

File: python/paddle/v2/data.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        items[i % size] = item
+    i += 1  # i is the total size.
+    i %= size
+    if i == 0:
+        yield items
+    else:
+        yield items[:min(i + 1, size)]
+
+
+class IDataPool(object):
```
Comment (reyoung) [inline python/paddle/v2/data.py]: @jacquesqiao @wangkuiyi 

Please review this interface. 

这个是Paddle API返回数据的接口。只有两个函数。另外，其实在Python里面，没有『接口』这个概念。这里写一下只是为了code清晰一点。

File: python/paddle/v2/data.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        if self.__idx__ >= len(self.__pool__):
+            raise StopIteration()
+
+        begin = self.__idx__
+        end = min(self.__idx__ + self.__batch_size__, len(self.__pool__))
+        self.__idx__ = end
+        return self.__pool__[begin:end]
+
+
+def create_data_pool(file_reader,
```
Comment (wangkuiyi) [inline python/paddle/v2/data.py]: 不同类型的data pool可能有各自不同的配置参数，不需要统一成同一组。我理解，用户自己调用 InMemoryDataPool 的构造函数创建一个 pool 就行，不需要 create_data_pool 这样一个函数。这样有违最简化的原则。

File: python/paddle/v2/layers.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+# limitations under the License.
+
+import paddle.trainer_config_helpers as conf_helps
+from paddle.trainer_config_helpers.config_parser_utils import \
+    parse_network_config as __parse__
+from paddle.trainer_config_helpers.default_decorators import wrap_name_default
+import collections
+
+
+class Layer(object):
```
Comment (wangkuiyi) [inline python/paddle/v2/layers.py]: 这里应该有一些注释，说明Layer应该怎么被使用。

之前 @hedaoyuan 在 review Error（Status）那个PR的时候提醒过需要预先定义语法（syntax）。我觉得非常有道理。

File: demo/mnist/api_train.py [hunk 17]
```diff
@@ -6,199 +6,76 @@
...
-        self.data = list(generator)
-        self.batch_size = batch_size
+    model.rand_parameter()
 
-    def __call__(self):
-        random.shuffle(self.data)
-        for offset in xrange(0, len(self.data), self.batch_size):
-            limit = min(offset + self.batch_size, len(self.data))
-            yield self.data[offset:limit]
+    batch_evaluator = model.make_evaluator()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 这两个evaluator看上去比较confusing。这里没有参数，所以看上去两个evaluator的功能是一样的。后面调用了.start和 .finish，但是也看不出来具体在干什么。根据名字（evaluator），猜测是用某种test data来评测模型，但是也看不出来用的是什么test data？

File: demo/mnist/api_train.py [hunk 18]
```diff
@@ -6,199 +6,76 @@
...
+        file_list=['./data/raw_data/test'],
+        model=model,
+        batch_size=128,
+        shuffle=False)
 
-def input_order_converter(generator):
-    for each_item in generator:
-        yield each_item['pixel'], each_item['label']
+    # Training process.
+    model.start()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: model不能start吧？这里想做的是 model.start_training?  在training开始之前需要做什么呢？

File: demo/mnist/api_train.py [hunk 19]
```diff
@@ -6,199 +6,76 @@
...
 
+    for pass_id in xrange(2):
+        model.start_pass()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: 我们之前讨论过，不应该由用户控制pass这个概念，因为pass是由数据集合大小决定的。我记得之前讨论过，可以写成如下方式：

```
for minibatch, last_batch in enumerate(training_data):
    ....
    if last_batch:
        log("Print model quality: %f", evaluate(model, testing_data))
    ....
```

File: demo/mnist/api_train.py [hunk 20]
```diff
@@ -6,199 +6,76 @@
...
-            value = value.copyToNumpyArray()
-
-            # Here, we could save parameter to every where you want
-            print each_param.getName(), value
 
-        updater.finishPass()
+        model.finish_pass()
 
-    m.finish()
+    model.finish()
```
Comment (wangkuiyi) [inline demo/mnist/api_train.py]: model.finish 要做什么呢？

**Comment by reyoung** [comment]
Closed because this PR's job is done by ourselves in last month.
